include/random_walks/shake_and_bake_walk.hpp

```diff
@@ -0,0 +1,147 @@
+// VolEsti (volume computation and sampling library)
+
+// Copyright (c) 
+
+// Licensed under GNU LGPL.3, see LICENCE file
+
+
+#ifndef RANDOM_WALKS_SHAKE_AND_BAKE_WALK_HPP
+#define RANDOM_WALKS_SHAKE_AND_BAKE_WALK_HPP
+
+#include <Eigen/Eigen>
+#include <limits>
+#include <cmath>
+#include "sampling/sphere.hpp"
+#include "convex_bodies/hpolytope.h"
+
+struct SBWalk
+{
+    enum Mode { Original, Limping, Running };
+
+    template <typename Polytope, typename RandomNumberGenerator>
+    struct Walk
+    {
+        using Point = typename Polytope::PointType;
+        using NT    = typename Point::FT;
+
+        Mode mode_{Mode::Original};
+
+        template <typename GenericPolytope>
+        Walk(GenericPolytope& P, Point const& p0, RandomNumberGenerator& rng, Mode m = Mode::Original)
+            : mode_{m}
+        {
+            initialize(P, p0);
+        }
+
+        template <typename GenericPolytope>
+        inline void apply(GenericPolytope const& P,
+                          Point& out_p,
+                          unsigned int const walk_len,
+                          RandomNumberGenerator& rng)
+        {
+            const NT eps = NT(1e-10);
+
+            for (unsigned t = 0; t < walk_len; ++t)
+            {
+                Point v = GetDirection<Point>::apply(P.dimension(), rng);
+
+                NT v_norm = v.getCoefficients().norm();
+                if (v_norm > eps) {
+                    for (std::size_t j = 0; j < dim_; ++j)
+                        v.set_coord(j, v[j] / v_norm);
+                }
+
+                NT dot_k = A_row_k_.dot(v.getCoefficients());
+                if (dot_k > NT(0)) {
+                    for (size_t j = 0; j < dim_; ++j)
+                        v.set_coord(j, -v[j]);
+                    dot_k = -dot_k;
+                }
+
+                int m_fac = static_cast<int>(P.num_of_hyperplanes());
+                Eigen::Matrix<NT, Eigen::Dynamic, 1> Ar(m_fac), Av(m_fac);
+                struct UP { NT inner_vi_ak; int facet_prev; };
+                UP params;
+
+                auto result = P.line_first_positive_intersect(p_, v, Ar, Av, params);
+                NT lambda_hit = result.first;
+                int r = result.second;
+
+                if (!std::isfinite(lambda_hit) || lambda_hit <= NT(0) || lambda_hit > NT(1e3) || r < 0)
+                    continue;
+
+                Point y = p_;
```

Comment:
Why do we need `y`? Could we implement Running SB without this additional variable?